File: internal/web/static/js/openfsd/airport-editor/format-apt.js
/**
 * Format Airport → TWRTrainer-compatible .apt text.
 * Port of pkg/twrfiles.FormatAPT — normative Format contract.
 */

import {
  DefaultRegistration,
  DEFAULT_PATTERN_SIZE,
  DEFAULT_INIT_CLIMB_PROPS,
  DEFAULT_INIT_CLIMB_JETS,
  SurfaceParking,
  SurfaceRunway,
  SurfaceTaxiway,
  SurfaceHold,
} from './model.js';

/**
 * Serialize an Airport to .apt text (LF, trailing newline, all headers).
 * @param {import('./model.js').Airport} a
 * @returns {string}
 */
export function formatAPT(a) {
  const icao = String(a?.icao ?? '').toUpperCase();
  let patternSize = a?.patternSize ?? 0;
  if (patternSize === 0) patternSize = DEFAULT_PATTERN_SIZE;
  let initProps = a?.initClimbProps ?? 0;
  if (initProps === 0) initProps = DEFAULT_INIT_CLIMB_PROPS;
  let initJets = a?.initClimbJets ?? 0;
  if (initJets === 0) initJets = DEFAULT_INIT_CLIMB_JETS;
  let reg = a?.registration ?? '';
  if (reg === '') reg = DefaultRegistration;

  const parts = [];
  parts.push(`icao=${icao}`);
  parts.push(`magnetic variation=${formatHeaderFloat(a?.magVar ?? 0)}`);
  parts.push(`field elevation=${formatHeaderFloat(a?.fieldElev ?? 0)}`);
  parts.push(`pattern elevation=${formatHeaderFloat(a?.patternElev ?? 0)}`);
  parts.push(`pattern size=${formatHeaderFloat(patternSize)}`);
  parts.push(`initial climb props=${formatHeaderFloat(initProps)}`);
  parts.push(`initial climb jets=${formatHeaderFloat(initJets)}`);
  parts.push(`jet airlines=${a?.jetAirlines ?? ''}`);
  parts.push(`turboprop airlines=${a?.turboAirlines ?? ''}`);
  parts.push(`registration=${reg}`);
  // Blank line after headers.
  parts.push('');

  const surfaces = Array.isArray(a?.surfaces) ? a.surfaces : [];
  for (let i = 0; i < surfaces.length; i++) {
    if (i > 0) parts.push('');
    writeSurface(parts, surfaces[i]);
  }

  // parts includes a trailing empty line body after headers so empty airports
  // become "...registration=N\n\n". Always terminate with LF (Go contract).
  return `${parts.join('\n')}\n`;
}

/**
 * @param {string[]} parts
 * @param {import('./model.js').Surface} s
 */
function writeSurface(parts, s) {
  switch (s.kind) {
    case SurfaceRunway: {
      let name = s.name;
      if ((!name || name === '') && (s.rwyA || s.rwyB)) {
        name = `${s.rwyA}/${s.rwyB}`;
      }
      parts.push(`[RUNWAY ${name}]`);
      const da = Math.trunc(s.dispA ?? 0);
      const db = Math.trunc(s.dispB ?? 0);
      parts.push(`displaced threshold=${da}/${db}`);
      // Match parse default (left when omitted). Intentionally not Go's zero-value
      // bool (false → right): hand-built partial surfaces without turnoffLeft
      // should still emit the TWRTrainer/parse default of left.
      const turnoffLeft = s.turnoffLeft ?? true;
      parts.push(turnoffLeft ? 'turnoff=left' : 'turnoff=right');
      break;
    }
    case SurfaceParking:
      parts.push(`[PARKING ${s.name}]`);
      break;
    case SurfaceTaxiway:
      parts.push(`[TAXIWAY ${s.name}]`);
      break;
    case SurfaceHold:
      parts.push(`[HOLD ${s.name}]`);
      break;
    default:
      parts.push(`[${s.kind} ${s.name}]`);
      break;
  }
  const pts = Array.isArray(s.points) ? s.points : [];
  for (const p of pts) {
    parts.push(`${formatCoord(p.lat)} ${formatCoord(p.lon)}`);
  }
}

/**
 * Always six fractional digits (Go %.6f). Non-finite → 0.000000 (defensive;
 * callers should validate coords; matches formatHeaderFloat finite guard).
 * @param {number} v
 */
export function formatCoord(v) {
  const n = Number(v);
  if (!Number.isFinite(n)) return '0.000000';
  return n.toFixed(6);
}

/**
 * Compact deterministic header float (Go strconv.FormatFloat 'f' -1).
 * @param {number} v
 */
export function formatHeaderFloat(v) {
  const n = Number(v);
  if (!Number.isFinite(n)) return '0';
  if (Object.is(n, -0) || n === 0) return '0';
  if (Number.isInteger(n) || Math.trunc(n) === n) {
    return String(Math.trunc(n));
  }
  // Prefer shortest decimal without scientific notation for airport ranges.
  let s = String(n);
  if (/e/i.test(s)) {
    s = n.toFixed(16).replace(/(\.\d*?)0+$/, '$1').replace(/\.$/, '');
  }
  return s;
}

File: pkg/twrfiles/format_apt.go
package twrfiles

import (
	"fmt"
	"strconv"
	"strings"
)

// FormatAPT serializes an Airport to TWRTrainer-compatible .apt text.
//
// Normative contract (docs/design/apt-air-editor.md):
//   - Always emit all ten header keys in fixed order (parse defaults when zero/empty)
//   - One blank line after headers; one blank line between surfaces
//   - Surfaces in Airport.Surfaces slice order (not re-bucketed by kind)
//   - RUNWAY always emits displaced threshold and turnoff
//   - Coordinates always %.6f %.6f; LF only; trailing newline
func FormatAPT(a Airport) string {
	var b strings.Builder

	icao := strings.ToUpper(a.ICAO)
	patternSize := a.PatternSize
	if patternSize == 0 {
		patternSize = defaultPatternSize
	}
	initProps := a.InitClimbProps
	if initProps == 0 {
		initProps = defaultInitClimbProps
	}
	initJets := a.InitClimbJets
	if initJets == 0 {
		initJets = defaultInitClimbJets
	}
	reg := a.Registration
	if reg == "" {
		reg = DefaultRegistration
	}

	b.WriteString("icao=")
	b.WriteString(icao)
	b.WriteByte('\n')

	b.WriteString("magnetic variation=")
	b.WriteString(formatHeaderFloat(a.MagVar))
	b.WriteByte('\n')

	b.WriteString("field elevation=")
	b.WriteString(formatHeaderFloat(a.FieldElev))
	b.WriteByte('\n')

	b.WriteString("pattern elevation=")
	b.WriteString(formatHeaderFloat(a.PatternElev))
	b.WriteByte('\n')

	b.WriteString("pattern size=")
	b.WriteString(formatHeaderFloat(patternSize))
	b.WriteByte('\n')

	b.WriteString("initial climb props=")
	b.WriteString(formatHeaderFloat(initProps))
	b.WriteByte('\n')

	b.WriteString("initial climb jets=")
	b.WriteString(formatHeaderFloat(initJets))
	b.WriteByte('\n')

	b.WriteString("jet airlines=")
	b.WriteString(a.JetAirlines)
	b.WriteByte('\n')

	b.WriteString("turboprop airlines=")
	b.WriteString(a.TurboAirlines)
	b.WriteByte('\n')

	b.WriteString("registration=")
	b.WriteString(reg)
	b.WriteByte('\n')

	// Blank line after headers (also provides trailing newline when no surfaces).
	b.WriteByte('\n')

	for i, s := range a.Surfaces {
		if i > 0 {
			b.WriteByte('\n')
		}
		writeSurface(&b, s)
	}

	return b.String()
}

func writeSurface(b *strings.Builder, s Surface) {
	switch s.Kind {
	case SurfaceRunway:
		name := s.Name
		if name == "" && (s.RwyA != "" || s.RwyB != "") {
			name = s.RwyA + "/" + s.RwyB
		}
		b.WriteString("[RUNWAY ")
		b.WriteString(name)
		b.WriteString("]\n")
		b.WriteString("displaced threshold=")
		b.WriteString(strconv.FormatInt(int64(s.DispA), 10))
		b.WriteByte('/')
		b.WriteString(strconv.FormatInt(int64(s.DispB), 10))
		b.WriteByte('\n')
		if s.TurnoffLeft {
			b.WriteString("turnoff=left\n")
		} else {
			b.WriteString("turnoff=right\n")
		}
	case SurfaceParking:
		b.WriteString("[PARKING ")
		b.WriteString(s.Name)
		b.WriteString("]\n")
	case SurfaceTaxiway:
		b.WriteString("[TAXIWAY ")
		b.WriteString(s.Name)
		b.WriteString("]\n")
	case SurfaceHold:
		b.WriteString("[HOLD ")
		b.WriteString(s.Name)
		b.WriteString("]\n")
	default:
		// Unknown kind: still emit a section so round-trip retains data order.
		b.WriteString("[")
		b.WriteString(s.Kind)
		b.WriteByte(' ')
		b.WriteString(s.Name)
		b.WriteString("]\n")
	}

	for _, p := range s.Points {
		b.WriteString(fmt.Sprintf("%.6f %.6f\n", p.Lat, p.Lon))
	}
}

// formatHeaderFloat emits a compact deterministic representation of a header
// numeric field (no unnecessary trailing zeros).
func formatHeaderFloat(v float64) string {
	return strconv.FormatFloat(v, 'f', -1, 64)
}
